python/cudf/cudf/core/udf/groupby_function.py

```diff
@@ -0,0 +1,851 @@
+
+import cupy as cp
+import numba
+import numpy as np
+from numba import cuda, types
+from numba.core import cgutils
+from numba.core.extending import (
+    lower_builtin,
+    make_attribute_wrapper,
+    models,
+    register_model,
+    type_callable,
+    typeof_impl,
+)
+from numba.core.typing import signature as nb_signature
+from numba.core.typing.templates import AbstractTemplate, AttributeTemplate
+from numba.cuda.cudadecl import registry as cuda_registry
+from numba.cuda.cudadrv.devices import get_context
+from numba.cuda.cudaimpl import lower as cuda_lower
+from numba.np import numpy_support
+from numba.types import Record
+
+from cudf.core.column import as_column
+from cudf.core.udf.templates import (
+    group_initializer_template,
+    groupby_apply_kernel_template,
+)
+from cudf.core.udf.utils import (
+    _all_dtypes_from_frame,
+    _compile_or_get,
+    _get_kernel_groupby_apply,
+    _get_udf_return_type,
+    _supported_cols_from_frame,
+    _supported_dtypes_from_frame,
+)
+from cudf.utils.utils import _cudf_nvtx_annotate
+
+# Disable occupancy warnings to avoid polluting output when there are few
+# groups.
+numba.config.CUDA_LOW_OCCUPANCY_WARNINGS = 0
+
+
+class Group(object):
+    def __init__(self, group_data, size, index, dtype, index_dtype):
+        self.group_data = group_data
+        self.size = size
+        self.index = index
+        self.dtype = dtype
+        self.index_dtype = index_dtype
+
+
+class GroupType(numba.types.Type):
+    def __init__(self, group_scalar_type, index_type=types.int64):
+        self.group_scalar_type = group_scalar_type
+        self.index_type = index_type
+        self.group_data_type = types.CPointer(group_scalar_type)
+        self.size_type = types.int64
+        self.group_index_type = types.CPointer(index_type)
+        super().__init__(
+            name=f"Group({self.group_scalar_type}, {self.index_type})"
+        )
+
+
+@typeof_impl.register(Group)
+def typeof_group(val, c):
+    return GroupType(
+        numba.np.numpy_support.from_dtype(val.dtype),
+        numba.np.numpy_support.from_dtype(val.index_dtype),
+    )  # converting from numpy type to numba type
+
+
+@type_callable(Group)
+def type_group(context):
+    def typer(group_data, size, index):
+        if (
+            isinstance(group_data, types.Array)
+            and isinstance(size, types.Integer)
+            and isinstance(index, types.Array)
+        ):
+            return GroupType(group_data.dtype, index.dtype)
+
+    return typer
+
+
+@register_model(GroupType)
+class GroupModel(models.StructModel):
+    def __init__(
+        self, dmm, fe_type
+    ):  # fe_type is fully instantiated group type
+        members = [
+            ("group_data", types.CPointer(fe_type.group_scalar_type)),
+            ("size", types.int64),
+            ("index", types.CPointer(fe_type.index_type)),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+my_max_int32 = cuda.declare_device(
```

Comment:
We should do all of this programmatically, rather than manually declaring every signature.

Reply:
Can we store these in a dict and generate them programmatically / obtain them by key that way? Something like

```python
cuda_functions = {}
cuda_functions['max'] = {}
for ty in supported_types:
    cuda_functions['max'][ty] = cuda.declare_device(
        f"BlockMax_{str(ty)}", 
        ty(types.CPointer(ty), 
        types.int64
    )
```
Later on in the lowering for `Max`, key into this dict to obtain the correct function based on `sig.return_type`.

Reply:
@brandon-b-miller can you explain how to do this programmatically but still using the wrapper function? I know that we cannot directly call the cuda.declare_device from the lowering and have to go through the wrapper function. I'm just not sure how to do it programmatically if we're to use the wrapper function.
`def call_my_max_int64(data, size):
    return my_max_int64(data, size)`

Reply:
Unfortunately using dictionary to programmatically generate the function gave me this error
```
Failed in nopython mode pipeline (step: nopython frontend)
Untyped global name 'cuda_functions': Cannot determine Numba type of <class 'dict'>

File "../python/cudf/cudf/core/udf/groupby_function.py", line 187:
def call_mean_int64(data, size): 
    return cuda_functions['mean']['int64'](data, size)
    ^

During: lowering "$10call_method.4 = call $8load_method.3(func=$8load_method.3, args=[], kws=(), vararg=None, varkwarg=None, targe
t=None)" at benchmark.py (46)
```